python/pyspark/sql/types.py

```diff
@@ -997,12 +997,13 @@ def _restore_object(dataType, obj):
     # same object in most cases.
     k = id(dataType)
     cls = _cached_cls.get(k)
-    if cls is None:
+    if cls is None or cls.__datatype is not dataType:
```

Comment:
Do we need this check for correctness?  If I understand the rest of this patch correctly, won't the write to `cls.__datatype` down on line 1006 mean that we'll retain a strong reference to the DataType as long as the Row class is still referenced, preventing the re-use of the DataType's object id?

If we expect this check to never return true, should we turn it into an assertion or log a warning if it's true?